### DISCUSSION TITLE
Why is a programmatic API not supported?

### DISCUSSION THREAD

**Discussion by matt-way** [npm/feedback#598] [CLI] [answered]
I need to bundle npm with an app, so that users can do things like install modules (they might not have npm installed). I'm currently using an older version of npm, and its programmatic api works fairly well. I wanted to know why it has been disabled with recent versions though? Is it due to design mistakes, or is there some technical hurdle I'm overlooking? I would have expected a cli layer to simply utilise the exact same api that could be exposed to consumers.

Is the ability to import/require npm on the roadmap?

**Comment by darcyclarke** [comment, marked as answer]
The programmatic usage of npm has never really been supported (ref. https://github.com/npm/npm/issues/8283#issuecomment-103718763) but we do our best to compartmentalize all the various functionality of the CLI into dependencies which do have programmatic APIs (ex. https://github.com/npm/pacote https://github.com/npm/arborist https://github.com/npm/npm-registry-fetch https://github.com/npm/libnpmpublish etc. etc.). You should be able to do anything the CLI does by pulling in our same dependencies.

**Reply by isaacs** [reply]
Also, if and when we re-enable a programmatic JavaScript API (ie, a way to do `require('npm').doStuff()` vs `proc = spawn('npm', ['do', 'stuff'])`), it'll be radically different and much more organized and well-documented than it has been in the past.  In order to get there, we have to shift things around quite a lot, drop `npm.load()`, and so on.  If the programmatic API is removed for at least one semver major version, we can do all that behind a wall, without making any user-facing breaking changes, without having to maintain a tricky compatibility layer, and with the ability to iterate without being committed to a programmatic API we may find is not ideal.

You _can_ use `@npmcli/arborist`, `pacote`, `@npmcli/config`, and so on, to do much of what npm does.  But if your goal is to replicate npm's behavior _exactly_ as it would be on the command line, then the only way to really do that for the foreseeable future is by shelling out to a child process.

**Reply by matt-way** [reply]
Thanks for the reply. Temporary walling makes sense. I'll look into rolling my own use of the cli deps for now. +1 for a future api though.

**Comment by SDAdham** [comment]
We also need to use npm programmatically, pls support it.
